src/core/Tools.cpp

```diff
@@ -39,6 +40,39 @@
 #include <time.h> // for nanosleep()
 #endif
 
+namespace
+{
+
+    bool hexDecodeNibble(uchar* result, char c)
+    {
+        if ('0' <= c && c <= '9') {
+            *result = c - '0';
+            return true;
+        } else if ('a' <= c && c <= 'f') {
+            *result = c - 'a' + 10;
+            return true;
+        } else if ('A' <= c && c <= 'F') {
+            *result = c - 'A' + 10;
+            return true;
+        } else {
+            std::cerr << "failed to hex-decode bad char '" << c << "'" << std::endl;
```

Comment:
Do not use `std::cerr`. Use `TextStream` if you must or `qWarning()` if it's not something the user should ever see. In fact, an assertion would be appropriate here.